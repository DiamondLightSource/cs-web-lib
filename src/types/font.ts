import { CSSProperties } from "react";

// assigned int values starting from zero
// also functions as a type
export enum FontStyle {
  Regular,
  Bold,
  Italic,
  BoldItalic
}

export class Font {
  // If no number is provided inherit from CSS.
  private size?: number;
  private style: FontStyle;
  private typeface: string;
  // TODO: name is currently unused
  private name?: string;

  public constructor(
    size?: number,
    style?: FontStyle,
    typeface?: string,
    name?: string
  ) {
    this.typeface = typeface ?? "Liberation sans";
    this.style = style ?? FontStyle.Regular;
    this.size = size;
    this.name = name;
  }

  /**
   * Returns a CSSProperties object containing all parameters defined by the
   * user and default properties for optional parameters not input by the user
   */
  public css(): CSSProperties {
    // Convert from px to REM by dividing pixels by default browser font size 16px
    // If a user has manually adjusted their browser font size, this might not be scaled
    // correctly, but given bob screens are pixel placed we shouldn't expect this to scale
    const fontSize = this.size ? `${this.size / 16}rem` : undefined;
    const fontWeight =
      this.style === FontStyle.Bold || this.style === FontStyle.BoldItalic
        ? "bold"
        : "normal";
    const fontStyle =
      this.style === FontStyle.Italic || this.style === FontStyle.BoldItalic
        ? "italic"
        : "normal";
    const style: CSSProperties = {
      // Fall back to sans-serif.
      fontFamily: `${this.typeface},sans-serif`,
      fontWeight: fontWeight,
      fontStyle: fontStyle
    };
    if (fontSize) {
      style.fontSize = fontSize;
    }
    return style;
  }
}
