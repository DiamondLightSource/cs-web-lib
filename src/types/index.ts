export { AbsolutePosition } from "./position";
export { RelativePosition } from "./position";
export { Color } from "./color";
export { PV } from "./pv";
export { Font, FontStyle } from "./font";
export { Border, BorderStyle } from "./border";
