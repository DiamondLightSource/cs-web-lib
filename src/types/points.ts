export class Point {
  public x: number;
  public y: number;

  public constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }
}

export class Points {
  public values: Array<Point>;

  public constructor(values: Array<Point>) {
    this.values = values;
  }
}
