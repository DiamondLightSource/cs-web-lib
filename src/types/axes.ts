import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public index: number;
  public autoScale?: boolean;
  public autoScaleThreshold?: number;
  public axisColor?: Color;
  public axisTitle?: string;
  public showGrid?: boolean;
  public gridColor?: Color;
  public dashGridLine?: boolean;
  public timeFormat?: number;
  public scaleFormat?: string;
  public scaleFont?: Font;
  public titleFont?: Font;
  public visible?: boolean;
  public logScale?: boolean;
  public leftBottomSide?: boolean;
  public maximum?: number;
  public minimum?: number;
  public yAxis?: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor(idx: number) {
    this.index = idx;
    this.autoScale = true;
    this.autoScaleThreshold = 0.95;
    this.axisColor = new Color("rgb(0, 0, 0");
    this.axisTitle = "";
    this.showGrid = false;
    this.gridColor = new Color("rgb(0, 0, 0");
    this.dashGridLine = true;
    this.timeFormat = 0;
    this.scaleFormat = "";
    this.scaleFont = new Font();
    this.titleFont = new Font(10, FontStyle.Bold);
    this.visible = true;
    this.logScale = false;
    this.leftBottomSide = true;
    this.minimum = 0;
    this.maximum = 100;
    // If axis has index 1 or above, assume y axis
    this.index > 0 ? (this.yAxis = true) : (this.yAxis = false);
  }
}

export class Axes {
  public count: number;
  public axisOptions: Axis[];

  public constructor(count: number, axes: Axis[]) {
    if (count !== axes.length) {
      throw new Error(
        `Count ${count} is not equal to number of axes ${axes.length}`
      );
    }
    this.count = count;
    this.axisOptions = axes;
  }
}
