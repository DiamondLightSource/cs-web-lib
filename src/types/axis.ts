import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public color: Color;
  public title: string;
  public showGrid: boolean;
  public visible: boolean;
  public logScale: boolean;
  public autoscale: boolean;
  public maximum: number;
  public minimum: number;
  public titleFont: Font;
  public scaleFont: Font;
  public onRight: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor({
    xAxis = false,
    color = new Color("rgb(0, 0, 0"),
    title = "",
    showGrid = false,
    visible = true,
    logScale = false,
    autoscale = false,
    minimum = 0,
    maximum = 100,
    titleFont = new Font(FontStyle.Bold),
    scaleFont = new Font(),
    onRight = false
  } = {}) {
    this.color = color;
    this.title = title || (xAxis ? "X" : "Y");
    this.showGrid = showGrid;
    this.visible = visible;
    this.logScale = logScale;
    this.autoscale = autoscale;
    this.minimum = minimum;
    this.maximum = maximum;
    this.titleFont = titleFont;
    this.scaleFont = scaleFont;
    this.onRight = onRight;
  }
}

export type Axes = Axis[];
