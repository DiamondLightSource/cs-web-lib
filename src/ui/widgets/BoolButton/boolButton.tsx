import React, { useEffect, useState, CSSProperties } from "react";
import { Widget } from "../widget";
import { PVComponent, PVWidgetPropType } from "../widgetProps";
import { registerWidget } from "../register";
import {
  InferWidgetProps,
  ColorPropOpt,
  IntPropOpt,
  PointsPropOpt,
  BoolPropOpt,
  StringPropOpt,
  FontPropOpt
} from "../propTypes";
import { Color } from "../../../types/color";
import classes from "./boolButton.module.css";
import { writePv } from "../../hooks/useSubscription";
import { DType } from "../../../types/dtypes";
import { WIDGET_DEFAULT_SIZES } from "../EmbeddedDisplay/bobParser";
import { Button as MuiButton, styled, useTheme } from "@mui/material";

// For HTML button, these are the sizes of the buffer on
// width and height. Must take into account when allocating
// maximum size for text
const BUTTON_BUFFER = { width: 12, height: 6 };

const BoolButtonProps = {
  pvName: StringPropOpt,
  height: IntPropOpt,
  width: IntPropOpt,
  onState: IntPropOpt,
  offState: IntPropOpt,
  onColor: ColorPropOpt,
  offColor: ColorPropOpt,
  onLabel: StringPropOpt,
  offLabel: StringPropOpt,
  squareButton: BoolPropOpt,
  foregroundColor: ColorPropOpt,
  backgroundColor: ColorPropOpt,
  points: PointsPropOpt,
  rotationAngle: IntPropOpt,
  effect3d: BoolPropOpt,
  showBooleanLabel: BoolPropOpt,
  showLed: BoolPropOpt,
  confirmMessage: StringPropOpt,
  labelsFromPv: BoolPropOpt,
  enabled: BoolPropOpt,
  font: FontPropOpt
};

const Button = styled(MuiButton)({
  "&.MuiButton-root": {
    display: "block",
    alignItems: "center",
    justifyContent: "center",
    height: "100%",
    width: "100%",
    minWidth: 0,
    minHeight: 0,
    padding: 0,
    overflow: "hidden",
    whiteSpace: "nowrap",
    wordBreak: "break-word",
    textTransform: "none"
  },
  "&.Mui-disabled": {
    cursor: "not-allowed",
    pointerEvents: "all !important"
  }
});

export type BoolButtonComponentProps = InferWidgetProps<
  typeof BoolButtonProps
> &
  PVComponent;

/**
 * Button that displays boolean value, and changes when clicked.
 * Currently no write to PV so value does not change, only button
 * appearance does
 * @param props
 */
export const BoolButtonComponent = (
  props: BoolButtonComponentProps
): JSX.Element => {
  const theme = useTheme();
  const {
    width = WIDGET_DEFAULT_SIZES["bool_button"][0],
    height = WIDGET_DEFAULT_SIZES["bool_button"][1],
    foregroundColor = theme.palette.primary.contrastText,
    backgroundColor = theme.palette.primary.main,
    pvName,
    value,
    onState = 1,
    offState = 0,
    onColor = Color.fromRgba(0, 255, 0),
    offColor = Color.fromRgba(0, 100, 0),
    squareButton = false,
    showBooleanLabel = true,
    showLed = true,
    labelsFromPv = false,
    enabled = true
  } = props;

  const font = props.font?.css() ?? theme.typography;

  // These could be overwritten by  PV labels
  let { onLabel = "On", offLabel = "Off" } = props;

  // Use labels from PV
  if (labelsFromPv) {
    if (value?.display.choices) {
      offLabel = value.display.choices[0];
      onLabel = value.display.choices[1];
    }
  }

  // Use useState for properties that change on click - text and color
  const [label, setLabel] = useState(showBooleanLabel ? offLabel : "");
  const [ledColor, setLedColor] = useState(offColor.toString());
  const doubleValue = value?.getDoubleValue();

  // Establish LED style
  const [ledStyle, ledDiameter] = showLed
    ? createLed(width, height, ledColor)
    : [{}, 0];
  if (squareButton) ledStyle["borderRadius"] = 0;

  const textStyle: CSSProperties = {
    // Ensure that text doesn't overflow from button
    maxWidth: width - ledDiameter - BUTTON_BUFFER.width,
    maxHeight: height - BUTTON_BUFFER.height
  };

  // Hide LED if it isn't visible
  if (showLed) ledStyle["visibility"] = "visible";

  // This is necessary in order to set the initial label value
  // after connection to PV established, as setState cannot be
  // established inside a conditional, or called in the main body
  // of the component as it causes too many re-renders error
  useEffect(() => {
    if (doubleValue === onState) {
      if (showBooleanLabel) setLabel(onLabel);
      setLedColor(onColor.toString());
    } else if (doubleValue === offState) {
      if (showBooleanLabel) setLabel(offLabel);
      setLedColor(offColor.toString());
    }
  }, [
    doubleValue,
    onState,
    onLabel,
    onColor,
    offState,
    offLabel,
    offColor,
    showBooleanLabel
  ]);

  function handleClick() {
    // Update button
    if (doubleValue === onState) {
      if (showBooleanLabel) setLabel(offLabel);
      setLedColor(offColor.toString());
    } else if (doubleValue === offState) {
      if (showBooleanLabel) setLabel(onLabel);
      setLedColor(onColor.toString());
    }
    // Write to PV
    if (pvName) {
      let newValue = offState;
      if (doubleValue === offState) newValue = onState;
      writePv(pvName, new DType({ doubleValue: newValue }));
    }
  }

  return (
    <>
      <Button
        variant="contained"
        onClick={handleClick}
        disabled={!enabled}
        sx={{
          fontFamily: font,
          color: foregroundColor.toString(),
          // If no LED, use on/off colours as background
          backgroundColor: showLed ? backgroundColor.toString() : ledColor
        }}
      >
        <span
          className={classes.LedAndText}
          style={{
            width: width - BUTTON_BUFFER.width,
            height: height - BUTTON_BUFFER.height
          }}
        >
          <span className={classes.Led} style={ledStyle} />
          <span className={classes.Text} style={textStyle}>
            {label}
          </span>
        </span>
      </Button>
    </>
  );
};

/**
 * Create an LED to display on BoolButton
 * @param width button width in px
 * @param height button height in px
 * @param color color of led
 * @returns CSSProperties for led
 */
export function createLed(
  width: number,
  height: number,
  color: string
): [CSSProperties, number] {
  // This is the same sizing as in Phoebus
  const size = Math.min(width, height);
  const ledRadius = size / 3.7;
  const ledDiameter = Math.round(ledRadius * 2);

  const ledStyle: CSSProperties = {
    width: ledDiameter,
    height: ledDiameter,
    backgroundColor: color,
    boxShadow: `inset ${ledDiameter / 4}px ${ledDiameter / 4}px ${
      ledDiameter * 0.4
    }px rgba(255,255,255,.5)`,
    visibility: "hidden"
  };

  return [ledStyle, ledDiameter];
}

const BoolButtonWidgetProps = {
  ...BoolButtonProps,
  ...PVWidgetPropType
};

export const BoolButton = (
  props: InferWidgetProps<typeof BoolButtonWidgetProps>
): JSX.Element => <Widget baseWidget={BoolButtonComponent} {...props} />;

registerWidget(BoolButton, BoolButtonWidgetProps, "boolbutton");
