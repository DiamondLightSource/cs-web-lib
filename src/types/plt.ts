import { Axes, Axis } from "./axis";
import { Color } from "./color";
import { Font } from "./font";
import { Trace } from "./trace";

export class Plt {
  public title: string;
  public axes: Axes;
  public pvlist: Trace[];
  public backgroundColor: Color;
  public foregroundColor: Color;
  public scroll: boolean;
  public scrollStep: number;
  public updatePeriod: number;
  public start: string;
  public end: string;
  public showGrid: boolean;
  public titleFont: Font;
  public scaleFont: Font;
  public labelFont: Font;
  public legendFont: Font;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property.
   */
  public constructor({
    title = "",
    axes = [new Axis()],
    pvlist = [new Trace()],
    background = Color.WHITE,
    foreground = Color.BLACK,
    scroll = true,
    grid = false,
    scrollStep = 5,
    updatePeriod = 1,
    titleFont = new Font(),
    labelFont = new Font(),
    legendFont = new Font(),
    scaleFont = new Font(),
    start = "1 minute",
    end = "now"
  } = {}) {
    this.backgroundColor = background;
    this.foregroundColor = foreground;
    this.title = title;
    this.scroll = scroll;
    this.scrollStep = scrollStep;
    this.axes = axes;
    this.pvlist = pvlist;
    this.updatePeriod = updatePeriod;
    this.showGrid = grid;
    this.start = start;
    this.end = end;
    this.titleFont = titleFont;
    this.scaleFont = scaleFont;
    this.labelFont = labelFont;
    this.legendFont = legendFont;
  }
}
