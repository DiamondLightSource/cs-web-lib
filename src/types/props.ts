import { Color } from "./color";
import { Font } from "./font";
import { MacroMap } from "./macros";
import { WidgetActions } from "../ui/widgets/widgetActions";
import { Border } from "./border";
import { Position } from "./position";
import { PV } from "./pv";
import { Trace } from "./trace";
import { Axes, Axis } from "./axis";
import { Points } from "./points";

export type GenericProp =
  | string
  | string[]
  | boolean
  | number
  | PV
  | Color
  | Font
  | Border
  | Position
  | Rule[]
  | MacroMap
  | WidgetActions
  | OpiFile
  | Trace[]
  | Axes
  | Axis
  | Points;

export interface Expression {
  boolExp: string;
  value: string;
  convertedValue?: GenericProp;
}

interface RulePV {
  pvName: PV;
  trigger: boolean;
}

export interface Rule {
  name: string;
  prop: string;
  outExp: boolean;
  pvs: RulePV[];
  expressions: Expression[];
}

export interface OpiFile {
  path: string;
  macros: MacroMap;
  defaultProtocol: string;
}
