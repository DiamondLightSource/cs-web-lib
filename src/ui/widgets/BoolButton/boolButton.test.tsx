import React from "react";
import { BoolButtonComponent } from "./boolButton";
import { fireEvent, render } from "@testing-library/react";
import { DType } from "../../../types/dtypes";
import { Color } from "../../../types";

const BoolButtonRenderer = (boolButtonProps: any): JSX.Element => {
  return <BoolButtonComponent {...boolButtonProps} />;
};

describe("<BoolButton />", (): void => {
  test("it renders a button with default values", (): void => {
    const boolButtonProps = {
      value: new DType({ doubleValue: 1 }),
      onState: 1,
      offState: 0
    };
    const { getByRole } = render(BoolButtonRenderer(boolButtonProps));
    const button = getByRole("button") as HTMLButtonElement;

    expect(button.textContent).toEqual("ON");
    expect(button.style.height).toEqual("50px");
    expect(button.style.width).toEqual("100px");
    expect(button.style.backgroundColor).toEqual("rgb(0, 255, 0)");
    expect(button.style.borderRadius).toEqual("50%");
  });

  test("it renders a button and overwrites default values", (): void => {
    const boolButtonProps = {
      value: new DType({ doubleValue: 1 }),
      width: 45,
      height: 20,
      onColor: Color.fromRgba(0, 235, 10),
      offColor: Color.fromRgba(0, 100, 0),
      onLabel: "Enabled",
      offLabel: "Disabled",
      squareButton: true,
      backgroundColor: Color.fromRgba(20, 20, 200),
      foregroundColor: Color.fromRgba(10, 60, 40),
      showBoolean: true,
      onState: 1,
      offState: 0
    };
    const { getByRole } = render(BoolButtonRenderer(boolButtonProps));
    const button = getByRole("button") as HTMLButtonElement;
    expect(button.textContent).toEqual("Enabled");
    expect(button.style.height).toEqual("20px");
    expect(button.style.width).toEqual("45px");
    expect(button.style.backgroundColor).toEqual("rgb(0, 235, 10)");
    expect(button.style.borderRadius).toEqual("");
  });

  test("no text if showboolean is false", (): void => {
    const boolButtonProps = {
      value: new DType({ doubleValue: 0 }),
      showBoolean: false,
      onState: 1,
      offState: 0
    };
    const { getByRole } = render(BoolButtonRenderer(boolButtonProps));
    const button = getByRole("button") as HTMLButtonElement;
    expect(button.textContent).toEqual("");
    expect(button.style.backgroundColor).toEqual("rgb(200, 200, 200)");
  });

  test("on click change text and colour ", async (): Promise<void> => {
    const boolButtonProps = {
      value: new DType({ doubleValue: 0 }),
      width: 45,
      height: 20,
      onColor: Color.fromRgba(0, 235, 10),
      offColor: Color.fromRgba(0, 100, 0),
      onLabel: "Enabled",
      offLabel: "Disabled",
      squareButton: true,
      backgroundColor: Color.fromRgba(20, 20, 200),
      foregroundColor: Color.fromRgba(10, 60, 40),
      showBoolean: true,
      onState: 1,
      offState: 0
    };
    const { getByRole } = render(BoolButtonRenderer(boolButtonProps));
    const button = getByRole("button") as HTMLButtonElement;

    // Original off values
    expect(button.textContent).toEqual("Disabled");
    expect(button.style.backgroundColor).toEqual("rgb(0, 100, 0)");

    // Click button to on
    fireEvent.click(button);

    expect(button.textContent).toEqual("Enabled");
    expect(button.style.backgroundColor).toEqual("rgb(0, 235, 10)");

    // Click button again to return to original
    fireEvent.click(button);
    expect(button.textContent).toEqual("Disabled");
    expect(button.style.backgroundColor).toEqual("rgb(0, 100, 0)");
  });
});
