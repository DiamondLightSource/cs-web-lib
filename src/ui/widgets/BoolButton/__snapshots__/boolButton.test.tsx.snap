// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<BoolButton /> > it renders a button with default values 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1rvqe9o-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      class="_LedAndText_69ddbc"
      style="width: 88px; height: 24px;"
    >
      <span
        class="_Led_69ddbc"
        style="width: 16px; height: 16px; background-color: rgb(0, 255, 0); box-shadow: inset 4px 4px 6.4px rgba(255,255,255,.5); visibility: visible;"
      />
      <span
        class="_Text_69ddbc"
        style="max-width: 72px; max-height: 24px;"
      >
        On
      </span>
    </span>
  </button>
</DocumentFragment>
`;

exports[`<BoolButton /> > it renders a button with led and overwrites default values 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-ry8voz-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      class="_LedAndText_69ddbc"
      style="width: 33px; height: 14px;"
    >
      <span
        class="_Led_69ddbc"
        style="width: 11px; height: 11px; background-color: rgb(0, 235, 10); box-shadow: inset 2.75px 2.75px 4.4px rgba(255,255,255,.5); visibility: visible; border-radius: 0;"
      />
      <span
        class="_Text_69ddbc"
        style="max-width: 22px; max-height: 14px;"
      >
        Enabled
      </span>
    </span>
  </button>
</DocumentFragment>
`;

exports[`<BoolButton /> > on click change text and led colour  1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-ry8voz-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      class="_LedAndText_69ddbc"
      style="width: 33px; height: 14px;"
    >
      <span
        class="_Led_69ddbc"
        style="width: 11px; height: 11px; background-color: rgb(0, 100, 0); box-shadow: inset 2.75px 2.75px 4.4px rgba(255,255,255,.5); visibility: visible; border-radius: 0;"
      />
      <span
        class="_Text_69ddbc"
        style="max-width: 22px; max-height: 14px;"
      >
        Disabled
      </span>
    </span>
  </button>
</DocumentFragment>
`;
